generator/main.cpp

```diff
@@ -125,53 +125,87 @@ namespace
+        dui.includePaths.push_back(QDir::toNativeSeparators(include).toStdString());
+      }
+      dui.defines.push_back("__cplusplus=1");
+      dui.defines.push_back("__STDC__");
+      dui.std = "c++17";
+      dui.removeComments = true;
 
-      QFile file(ppconfig);
+      QFile file(sourceFile);
       if (!file.open(QFile::ReadOnly))
       {
-        fprintf(stderr, "Preprocessor configuration file not found '%s'\n", ppconfig);
+        std::cerr << "Main file not found:" << sourceFile.toUtf8().constData() << std::endl;
         return false;
       }
 
       QByteArray ba = file.readAll();
       file.close();
-      preprocess.operator()(ba.constData(), ba.constData() + ba.size(), null_out);
 
-      foreach(QString
-      include, getIncludeDirectories(commandLineIncludes)) {
-        preprocess.push_include_path(QDir::toNativeSeparators(include).toStdString());
+      // Perform preprocessing
+      simplecpp::OutputList outputList;
+      std::vector<std::string> files;
+      simplecpp::TokenList* rawtokens = new simplecpp::TokenList(ba.constData(), ba.size(), files, {}, &outputList);
+      rawtokens->removeComments();
+      simplecpp::TokenList outputTokens(files);
+      std::map<std::string, simplecpp::TokenList*> filedata;
+      simplecpp::preprocess(outputTokens, *rawtokens, files, filedata, dui, &outputList);
+      simplecpp::cleanup(filedata);
+      delete rawtokens;
+      rawtokens = nullptr;
+
+      for (const simplecpp::Output& output : outputList) {
+        if (output.type == simplecpp::Output::MISSING_HEADER) {
+          // do not print these messages for now, we are not interested
+          continue;
+        }
+        std::cerr << output.location.file() << ':' << output.location.line << ": ";
+        switch (output.type) {
+        case simplecpp::Output::ERROR:
+          std::cerr << "#error: ";
+          break;
+        case simplecpp::Output::WARNING:
+          std::cerr << "#warning: ";
+          break;
+        case simplecpp::Output::MISSING_HEADER:
+          std::cerr << "missing header: ";
+          break;
+        case simplecpp::Output::INCLUDE_NESTED_TOO_DEEPLY:
+          std::cerr << "include nested too deeply: ";
+          break;
+        case simplecpp::Output::SYNTAX_ERROR:
+          std::cerr << "syntax error: ";
+          break;
+        case simplecpp::Output::PORTABILITY_BACKSLASH:
+          std::cerr << "portability: ";
+          break;
+        case simplecpp::Output::UNHANDLED_CHAR_ERROR:
+          std::cerr << "unhandled char error: ";
+          break;
+        case simplecpp::Output::EXPLICIT_INCLUDE_NOT_FOUND:
+          std::cerr << "explicit include not found: ";
+          break;
+        case simplecpp::Output::FILE_NOT_FOUND:
+          std::cerr << "file not found: ";
+          break;
+        case simplecpp::Output::DUI_ERROR:
+          std::cerr << "dui error: ";
+          break;
+        }
+        std::cerr << output.msg << std::endl;
       }
```

Comment:
Maybe mention that this is taken from simplecpp/main.cpp.

Reply:
Ok.